### ISSUE TITLE
make the communication fail-safe

### ISSUE THREAD

**Issue by Trolldemorted** [Frodo45127/rpfm#15]
If I understand #14 correctly, the communication between ui and bg thread may fail, if the user triggers multiple commands at once. Taken from #14:

Imagine we have the commands A0, A1 and B. All three return data, A1 will only be sent after A0 finishes, and A0 and B are sent by user interaction. 
Now
- the user initiates A0
- the A0 handler is in a tryrecv-loop, so the user can trigger new commands
- the user initiates B
The command channel contains: B -> A0
The bg->ui data channel contains: nothing

Now
- the bg thread handles A0, puts a0-data in the bg->ui channel
- the ui thread handles a0-data, and puts A1 into the ui->bg command channel
The command channel contains: A1 -> B
The bg->ui data channel contains: nothing

Now
- the bg thread handles B and puts b-data into the bg->ui channel
- the ui thread expects a1-data but gets b-data 💣 💥 

To address that, we must prevent overlapping access to the ui->bg (command) channel. How to archive that depends on what behaviour you want:

### 1. No command queue
Dead simple: Track channel-usage in a bool, only issue a new command when it is false, set it to true before issuing, set it to false after completion.
#### Pros:
- simple
- ui still feels responsive because resizing and menu traversing handling still works
- you can upgrade it to 2. later if you feel like it
#### Cons:
- users have to wait until a command completes before they can issue a new one

### 2. (Intelligent) command queue
Basically 1. with a dedicated queue attached. UI handlers enqueue commands into the queue, the UI thread takes a new command out of the queue once its predecessor has finished.
#### Pros:
- if a user performs an action n times, the backend will try to execute it n times
#### Cons:
- complicated
- if a user performs an action n times, the backend will try to execute it n times

**Comment by Frodo45127** [comment]
Ok, closing this as with the new client/server architecture coming in 5.0, it should be safe enough.
